hal/armv7m/exceptions.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * Exception handling
+ *
+ * Copyright 2017, 2023 Phoenix Systems
+ * Author: Pawel Pisarczyk, Jakub Sejdak
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+
+#include <hal/hal.h>
+
+#define SIZE_FPUCTX (16 * sizeof(u32))
+
+#ifdef CPU_IMXRT
+#define EXCRET_PSP 0xffffffed
+#else
+#define EXCRET_PSP 0xfffffffd
+#endif
+
+
+typedef struct {
+	u32 r0;
+	u32 r1;
+	u32 r2;
+	u32 r3;
+	u32 r12;
+	u32 lr;
+	u32 pc;
+	u32 psr;
+} cpu_hwContext_t;
+
+
+typedef struct _exc_context_t {
+	u32 psp;
+	u32 r4;
+	u32 r5;
+	u32 r6;
+	u32 r7;
+	u32 r8;
+	u32 r9;
+	u32 r10;
+	u32 r11;
+	u32 excret;
+
+	/* Saved by hardware */
+	cpu_hwContext_t mspctx;
+} exc_context_t;
+
+
+__attribute__((section(".noxip"))) void hal_exceptionsDumpContext(char *buff, exc_context_t *ctx, int n)
+{
+	static const char *const mnemonics[] = {
+		"0 #InitialSP", "1 #Reset", "2 #NMI", "3 #HardFault",
+		"4 #MemMgtFault", "5 #BusFault", "6 #UsageFault", "7 #",
+		"8 #", "9 #", "10 #", "11 #SVC",
+		"12 #Debug", "13 #", "14 #PendSV", "15 #SysTick"
+	};
+
+	cpu_hwContext_t *hwctx;
+	u32 msp = (u32)ctx + sizeof(*ctx);
+	u32 psp = ctx->psp;
+	const u32 fpucheck = (*(u32 *)0xe000ed88) & ((3uL << 20) | (3uL << 22));
+	char *ptr = buff;
+
+	n &= 0xf;
+
+	/* If we came from userspace HW ctx in on psp stack */
+	if (ctx->excret == EXCRET_PSP) {
+		hwctx = (void *)ctx->psp;
+		msp -= sizeof(cpu_hwContext_t);
+		psp += sizeof(cpu_hwContext_t);
+		if (fpucheck != 0) {
+			psp += SIZE_FPUCTX;
+		}
+	}
+	else {
+		hwctx = &ctx->mspctx;
+		if (fpucheck != 0) {
+			msp += SIZE_FPUCTX;
+		}
+	}
+
+	hal_strcpy(ptr, "\nException: ");
+	ptr += hal_strlen(ptr);
+
+	hal_strcpy(ptr, mnemonics[n]);
+	ptr += hal_strlen(ptr);
+
+	hal_strcpy(ptr, "\n");
+	ptr += hal_strlen(ptr);
+
+	ptr += hal_i2s(" r0=", ptr, hwctx->r0, 16, 1);
```

Comment:
```suggestion
	ptr += hal_i2s("\n r0=", ptr, hwctx->r0, 16, 1);
```
Calling `hal_strcpy()` and `hal_strlen()` for one char?

Reply:
Thanks, fixed.